provide/Services/MessageProtocolParser.cs

```diff
@@ -0,0 +1,89 @@
+using System.IO;
+using System.Text;
+
+public class Message
+{
+    public string OpCode { get => "BLINE"; } // up to 40 bits
+    public string Sender { get; set; } // up to 336 bits
+    public string Recipient { get; set; } // up to 336 bits
+    public string Shield { get; set; } // up to 336 bits
+    public string Identifier { get; set; } // up to 288 bits (i.e., UUIDv4 circuit/workflow identifier)
+    public string Signature { get; set; } // 512 bits
+    public string Type { get => "1"; }  // 1 bit
+    public byte[] Payload { get; set; } // arbitrary length
+}
+
+public static class MessageProtocolParser
+{
+    const int MessageReservedBitsLength = 512;
+    public static byte[] MarshalProtocolMessage(Message msg)
+    {
+        var reservedBits = new byte[MessageReservedBitsLength / 8];
+        var buffer = new byte[5 + 42 + 42 + 42 + 36 + 64 + 1 + reservedBits.Length + msg.Payload.Length];
+
+        using (MemoryStream ms = new MemoryStream(buffer))
+        {
+            using(BinaryWriter bw = new BinaryWriter(ms, Encoding.UTF8))
+            {
+                bw.Write(msg.OpCode.ToCharArray());
+
+                bw.Seek(5, SeekOrigin.Begin);
+                bw.Write(msg.Sender.ToCharArray());
+
+                bw.Seek(5 + 42, SeekOrigin.Begin);
+                bw.Write(msg.Recipient.ToCharArray());
+
+                bw.Seek(5 + 42 + 42, SeekOrigin.Begin);
+                bw.Write(msg.Shield.ToCharArray());
+
+                bw.Seek(5 + 42 + 42 + 42, SeekOrigin.Begin);
+                bw.Write(msg.Identifier.ToCharArray());
+
+                bw.Seek(5 + 42 + 42 + 42 + 36, SeekOrigin.Begin);
+                bw.Write(reservedBits);
+
+                bw.Seek(5 + 42 + 42 + 42 + 36 + reservedBits.Length, SeekOrigin.Begin);
+                bw.Write(msg.Signature.ToCharArray());
+
+                bw.Seek(5 + 42 + 42 + 42 + 36 + reservedBits.Length + 64, SeekOrigin.Begin);
+                bw.Write(msg.Type.ToCharArray());
+
+                bw.Seek(5 + 42 + 42 + 42 + 36 + reservedBits.Length + 64 + 1, SeekOrigin.Begin);
+                bw.Write(msg.Payload);
+            }
+        }
+    
+        return buffer;
+    }
+
+    public static Message UnmarshalProtocolMessage (byte[] msg)
+    {
+        var reservedSize = MessageReservedBitsLength / 8;
+
+        using (var ms = new MemoryStream(msg))
+        {
+            using (var br = new BinaryReader(ms, Encoding.UTF8))
+            {
+                var opCode = br.ReadBytes(5);
+                var sender = Encoding.UTF8.GetString(br.ReadBytes(42));
```

Comment:
i used unit test with some random strings to verify if unmarshaling will give back original message but didn't include it in test suite because data is random, do we have some message example with proper data to test it out?

Reply:
just create one based on the description of the baseline message in the data model page in conf